File: monty_hall_gen.py
import random
import tqdm
import argparse


def monty_hall_calc(doors: int, switch: bool) -> float:
	switch_val = int(switch)

	prob = switch_val/ (doors - 2)
	prob += 1
	prob /= doors
	return prob


def monty_hall_sim(switch: bool, tests: int, doors: int) -> float:
	i = 0
	hits = 0
	pbar = tqdm.tqdm(total=tests)

	
	while i < tests:
		doors_arr = [j for j in range(doors)]
		chosen_door = random.choice(doors_arr)
		correct_door = random.choice(doors_arr)

		doors_arr.remove(chosen_door)

		oppened_door_ind = random.randint(0, len(doors_arr)-1)
		oppened_door = doors_arr[oppened_door_ind]
		if correct_door == oppened_door:
			if oppened_door == doors_arr[-1]:
				oppened_door_ind -= 1
			else:
				oppened_door_ind += 1
			oppened_door = doors_arr[oppened_door_ind]
		doors_arr.remove(oppened_door)
		
		if switch:
			chosen_door = random.choice(doors_arr)
		
		if correct_door == chosen_door:
			hits += 1
		
		i += 1
		pbar.update(1)
	
	pbar.close()
	hits = round(hits / tests, 4)
	return hits


parser = argparse.ArgumentParser(prog='GenMontyHall_py', description='Monty hall simulator - Python version')

parser.add_argument('-t', '--tests', help='number of tests', type=int, default=100000000)
parser.add_argument('-d', '--doors', help='number of doors', type=int, default=3)

args = parser.parse_args()

tests = args.tests
doors = args.doors

print(f'Probability of winning with {doors} doors when switching: ', end='')
print(f'{monty_hall_calc(doors=doors, switch=True) * 100}%\n')

print(f'Simulating games with {doors} doors and switching...')
hits = monty_hall_sim(switch=True, tests=tests, doors=doors)
print(f'Percentage of games won: {hits * 100}%')

print(f'Probability of winning with {doors} doors but no switching: ', end='')
print(f'{monty_hall_calc(doors=doors, switch=False) * 100}%\n')

print(f'Simulating games with {doors} doors but no switching...')
hits = monty_hall_sim(switch=False, tests=tests, doors=doors)
print(f'Percentage of games won: {hits * 100}%')

File: monty_hall_gen.go
package main

import (
	"fmt"
	"github.com/peterbourgon/ff/v4"
	"github.com/peterbourgon/ff/v4/ffhelp"
	"github.com/schollz/progressbar/v3"
	"math/rand"
	"os"
)

func monty_hall_calc(doors int, switching bool) float32 {

	var switching_val float32 = 0
	if switching {
		switching_val = 1.0
	}

	prob := switching_val
	prob = prob / (float32(doors - 2))
	prob++
	prob = prob / float32(doors)
	return prob
}

func monty_hall_sim(test int, doors int, switching bool) (float32, error) {
	var hits int = 0

	var prog_bar = progressbar.Default(int64(test), "Simulating")

	for i := 0; i < test; i++ {
		var doors_arr []int

		for j := 0; j < doors; j++ {
			doors_arr = append(doors_arr, j)
		}

		var correct_door = doors_arr[rand.Intn(len(doors_arr))]
		var chosen_door = doors_arr[rand.Intn(len(doors_arr))]

		var doors_copy = append(doors_arr[:chosen_door], doors_arr[chosen_door+1:]...)

		var openned_door_ind = rand.Intn(len(doors_copy))
		var oppened_door = doors_copy[openned_door_ind]
		if oppened_door == correct_door {
			if oppened_door == doors_copy[len(doors_copy)-1] {
				openned_door_ind--
			} else {
				openned_door_ind++
			}
		}

		doors_copy = append(doors_copy[:openned_door_ind], doors_copy[openned_door_ind+1:]...)

		if switching {
			chosen_door = doors_copy[rand.Intn(len(doors_copy))]
		}

		if chosen_door == correct_door {
			hits++
		}

		err := prog_bar.Add(1)
		if err != nil {
			return 0, err
		}
	}

	err := prog_bar.Close()
	if err != nil {
		return 0, nil
	}

	return float32(hits), nil
}

func main() {
	flag_set := ff.NewFlagSet("GenMontyHall_go")
	var (
		doors = flag_set.Int('d', "doors", 3, "number of doors")
		tests = flag_set.Int('t', "tests", 10000000, "number of tests")
		help  = flag_set.Bool('h', "help", "help")
	)

	err := ff.Parse(flag_set, os.Args[1:])
	if err != nil {
		fmt.Println("Paring arguments failed: ", err)
	}

	if *help {
		fmt.Printf("%s\n", ffhelp.Flags(flag_set))
		os.Exit(0)
	}

	var switch_prob = monty_hall_calc(*doors, true) * 100
	fmt.Printf("Probability of winning when switching %.2f%%\n", switch_prob)

	switch_sim, err := monty_hall_sim(*tests, *doors, true)
	if err != nil {
		fmt.Printf("Error happened with progress bar, poor you %v\n", err)
		os.Exit(1)
	}
	switch_sim = switch_sim / float32(*tests) * 100
	fmt.Printf("Percentage of wins when switching %.2f\n", switch_sim)

	var non_switch_prob = monty_hall_calc(*doors, false) * 100
	fmt.Printf("Probability of winning when not switching %.2f%%\n", non_switch_prob)

	non_switch_sim, err := monty_hall_sim(*tests, *doors, false)
	if err != nil {
		fmt.Printf("Error happened with progress bar, poor you %v\n", err)
		os.Exit(1)
	}

	non_switch_sim = non_switch_sim / float32(*tests) * 100
	fmt.Printf("Percentage of wins when switching %.2f", non_switch_sim)
}
